src/hivpy/sexual_behaviour.py

```diff
@@ -40,9 +72,38 @@ def prob_transition(self, gender, age, i, j):
 
         return Probability
 
-    def num_short_term_partners(self, population):
-        population["short_term_partners"] = np.where(population['gender'] == 'male',
-                                                     self.short_term_partners_male
-                                                     [population['sex_behaviour']].rvs(),
-                                                     self.short_term_partners_female
-                                                     [population['sex_behaviour']].rvs())
+    def num_short_term_partners(self, population: pd.DataFrame):
+        for sex in SexType:
+            for g in SexBehaviours[sex]:
+                index = selector(population, sex=(operator.eq, sex),
+                                 sex_behaviour=(operator.eq, g.value))
+                population.loc[index, "num_partners"] = (
+                    self.short_term_partners[sex][g.value].rvs(size=sum(index)))
+
+    def update_sex_groups(self, population: pd.DataFrame):
+        """Determine changes to sexual behaviour groups.
+           Loops over sex, and behaviour groups within each sex.
+           Within each group it then loops over groups again to check all transition pairs (i,j)."""
+        pop_size = len(population)
+
+        for sex in SexType:
+            for i in SexBehaviours[sex]:
```

Comment:
I would be explicit about this before we get lost with i and j, for example:

```suggestion
            for old_behaviour in SexBehaviours[sex]:
```

Reply:
Will update this